### ISSUE TITLE
Sub interface method overload with base interface parameter is not checked for correctness

### ISSUE THREAD

**Issue by mwienand** [Microsoft/TypeScript#17980]
**TypeScript Version:**  2.4.2

**Code**

```ts
interface Base {
    can(x: number);
}

interface Special extends Base {
    can(x: number);
    can(x: Base);
}

class A implements Base {
    a: string;
    constructor() { this.a = "A" }
    can(x: number) {
    }
}

class B implements Special {
    b: string;
    constructor() { this.b = "B" }
    can(x: number);
    can(x: A);
    can(x: number | A) {
        // Why is this implementation sufficient?
        console.log("Given: " + x);
    }
}

let a = new A();
let b: Special = new B();
b.can(b);
```

**Expected behavior:**

The definition of class B should not be correct, because the can(x: Base) method is missing from B.

**Actual behavior:**

No errors. can(x: number | A) is called with a B. By writing "class X implements I", I want to ensure that the interface is correctly implemented. However, the missing method is not reported.

**Comment by aluanhaddad** [comment]
`B` is a perfectly valid implementation of `Special` because `A` is a perfectly valid implementation of `Base`.
It works for the same reason that
```ts
const b: Base = new A();
const s: Special = new Special();
s.can(b);
```
works.

**Comment by mwienand** [comment]
I do not understand that reasoning, because the following triggers a type error:

```
let a = new A();
let b = new B();
b.can(b); // type error
```

In general, the interface allows more objects than the implementation type, because the implementation type contains further properties and methods.

**Comment by mwienand** [comment]
Furthermore, the union type number | A should not accept B, however, that is what is passed to the method in the initial example.

**Comment by mhegazy** [comment]
you can find some explanation for this in https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant

**Comment by mwienand** [comment]
FYI, I implemented a tslint rule for collecting type information (interfaces and classes), and a second rule that creates failures if an implementation does not provide a method as is described in one of its interfaces, i.e. all interface methods need to be present in an implementation (with types matching the exact description in the interface).

Therefore, in the initial example, an error message would be present at the B#can(...) method, stating that an implementation is missing for Special#can(x: Base).

I still fail to understand why this check cannot be performed by TypeScript and why the current behaviour is considered correct.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
